### ISSUE TITLE
Questionable str_replace behavior

### ISSUE THREAD

**Issue by iakoder** [kalessil/phpinspectionsea#1129]
| Subject        | Details                                                                       |
| :------------- | :---------------------------------------------------------------------------- |
| Issue type     | Maybe false positive?                          |
| Plugin         | Php Inspections (EA Extended), 3.0.9 |
| Language level | PHP 7.0                        |

Was there a change in 3.0.9 regarding the signature of str_replace?  I just updated and started seeing repeated calls to str_replace start firing "new value type (array) is not in annotated types."

```
<?php
function f(string $arg) {
    $arg = str_replace("foo", "bar", $arg);
    // New value type (array) is not in annotated types
    $arg = str_replace("foo", "bar", $arg);
}
```

The signature (in PhpStorm 2018.3) is 
```
/**
 * Replace all occurrences of the search string with the replacement string
 * @link https://php.net/manual/en/function.str-replace.php
 * @param mixed $search <p>
 * The value being searched for, otherwise known as the needle.
 * An array may be used to designate multiple needles.
 * </p>
 * @param mixed $replace <p>
 * The replacement value that replaces found search
 * values. An array may be used to designate multiple replacements.
 * </p>
 * @param mixed $subject <p>
 * The string or array being searched and replaced on,
 * otherwise known as the haystack.
 * </p>
 * <p>
 * If subject is an array, then the search and
 * replace is performed with every entry of
 * subject, and the return value is an array as
 * well.
 * </p>
 * @param int $count [optional] If passed, this will hold the number of matched and replaced needles.
 * @return mixed This function returns a string or an array with the replaced values.
 * @since 4.0
 * @since 5.0
 */
function str_replace ($search, $replace, $subject, &$count = null) {}
```

So I'm curious where you're getting array from and even if you do know that the return type might be array it would be nice if you also applied the fact that the return type is only array if the subject is an array.

**Comment by kalessil** [comment]
Fixed!

**Comment by iakoder** [comment]
@kalessil Just ran 3.0.9.1.  Here are the cases it still found

```
function f2(string $inVal) {
    //  Change any double backslash to single backslash.
    $inVal = str_replace("foo", "bar", $inVal);

    $len = strlen($inVal);
    if ($inVal[0] === "'" && strrpos($inVal, "'") == $len - 1) {
        $inVal = (string)substr($inVal, 1, $len - 2);
    }

    // New value type (array) is not in annotated types
    $inVal = str_replace("foo", "bar", $inVal);

}

/**
 * @param string|float $arg
 */
function f3($arg) {
    // New value type (array) is not in annotated types
    $arg = str_replace("foo", "bar", $arg);
}

function f4(string $arg) {
    $a = [ 'key' => 'value' ];
    foreach ($a as $k => $v) {
        // New value type (array) is not in annotated types, probably same as f3()
        $arg = str_replace("foo", "bar", $k);
    }
}

function f5(string $arg) {
    // New value type (array) is not in annotated types
    $arg = substr_replace($arg, $this->exportfilters, 0, 10);
}


function f6(string $arg, bool $b1, bool $b2) {
    if ($b1) {
        $arg = str_replace("foo", "bar", $arg);
    }
    if ($b2) {
        // New value type (array) is not in annotated types
        $arg = str_replace("foo", "bar", $arg);
    }

}

```

**Comment by kalessil** [comment]
Snap, I'll keep fixing then.

**Comment by kalessil** [comment]
Fixed!
